File: Уровень2/Уровень2А/Задачи_на_префиксные_суммы/Количество_нулей_на_интервале.py
#https://algoprog.ru/material/p3313

if __name__ == "__main__":
    n = int(input())
    array = [*map(int, input().split())]
    
    cnt = [0 for _ in range(n)]
    cnt[0] = 1 if array[0] == 0 else 0
    for i in range(1, n):
        if array[i] == 0:
            cnt[i] = cnt[i-1] + 1
        else:
            cnt[i] = cnt[i-1]
                
    k = int(input())
    for i in range(k):
        left, right = map(int, input().split())
        left -= 1
        right -= 1
        if left > 0:
            print(cnt[right] - cnt[left-1])
        else:
            print(cnt[right])

File: Уровень2/Уровень2А/Задачи_на_префиксные_суммы/Количество_нулей_на_интервале.cpp
// https://algoprog.ru/material/p3313

#include <iostream>
#include <vector>
using namespace std;

int main() {
    int n, m, k;
    cin >> n >> m >> k;
    int array[n][m];
    for (auto i = 0; i < n; i++) {
        for (auto j = 0; j < m; j++) {
            int tmp;
            cin >> tmp;
            array[i][j] = tmp;
        }
    }
    int prefix[n][m];
    for (auto i = 0; i < n; i++) {
        for (auto j = 0; j < m; j++) {
            if (i == 0 and j == 0) {
                prefix[i][j] = array[i][j];
            } else if (i >= 1 and j == 0) {
                prefix[i][j] = array[i][j] + prefix[i-1][m-1];
            } else {
                prefix[i][j] = array[i][j] + prefix[i][j-1];
            }
        }
    }
    for (auto j = 0; j < k; j++) {
        int x1, y1, x2, y2;
        cin >> x1 >> y1 >> x2 >> y2;
        x1--;
        y1--;
        x2--;
        y2--;
        int cnt = 0;
        for (auto i = x1; i < x2 + 1; i++) {
            if (i == 0 and y1 == 0) {
                cnt += prefix[i][y2];
            } else if (i >= 1 and y1 == 0) {
                cnt += prefix[i][y2] - prefix[i-1][m-1];
            } else {
                cnt += prefix[i][y2] - prefix[i][y1-1];
            }
        }
        cout << cnt << endl;
    }
    return 0;
}
